File: choitaejin/week5/boj_2579.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let score = [];
let lineCnt = 0;
let n;

rl.on('line', (line) => {
    if (lineCnt === 0) {
        n = parseInt(line);
        lineCnt++;
    }
    else if (lineCnt <= n) {
        score.push(parseInt(line));
        lineCnt++;
        if(lineCnt > n) {
            rl.close();
        }
    }

}).on('close', () => { // [i][j] = i는 밟은 계단 수 - 1, j는 연속해서 밟은 계단 수 - 1
    let dp = Array(n).fill().map(() => [0, 0]);
    dp[0][0] = score[0]; // 첫 번째 계단 밟은 경우
    dp[0][1] = 0; // 첫번째 계단을 밟았는데 2개 연속해서 밟은 경우는 없으니까
    
    if (n > 1) {
        dp[1][0] = score[1]; // 첫 번째 계단 밟고 두번째
        dp[1][1] = score[0] + score[1]; // 두 번째 계단을 밟을 때 첫번째 밟고 두번째 연속으로 밟은 경우

        for (let i = 2; i < n; i++) {
            dp[i][0] = Math.max(dp[i-2][0], dp[i-2][1]) + score[i];
            dp[i][1] = dp[i-1][0] + score[i];
        }
    }

    console.log(Math.max(dp[n-1][0], dp[n-1][1]));

    process.exit();
});

File: leeyesol/week5/boj_2579.c
#include <stdio.h>

// 두 수 중 더 큰 값을 반환하는 함수
int max(int a, int b)
{
	return a > b ? a : b;
}

int main()
{
	int i, n;
	int input[301]; // 각 계단의 점수 저장
	int dp[301];    // dp[i]: i번째 계단까지 올라갔을 때의 최대 점수

	scanf("%d", &n); // 계단의 개수 입력

	i = 1;
	while (i <= n)
	{
		scanf("%d", &input[i]); // 계단 점수 입력 (1번 인덱스부터 시작)
		i++;
	}

	// 초기 조건 설정
	dp[1] = input[1];                  // 첫 번째 계단 점수
	dp[2] = input[1] + input[2];       // 첫 + 두 번째 계단 점수

	i = 3;
	while (i <= n)
	{
		// 두 가지 경우 중 큰 값을 선택
		// 1) i-2 → i
		// 2) i-3 → i-1 → i
		dp[i] = max(dp[i - 2] + input[i], dp[i - 3] + input[i - 1] + input[i]);
		i++;
	}

	printf("%d", dp[n]); // 마지막 계단까지의 최대 점수 출력
}
